[MERGE] Support two-factor authentication + API keys

xmo-odoo opened the pull request:
- Adds support for two-factor authentication (2FA) via time-based one-time-passwords (TOTP),
  compatible with mainstream authenticators like Google Auth, 1Password, etc.
- Add support for personal API key that can be used for non-interactive RPC scripting, including
  when 2FA is enabled (password-based non-interactive RPC auth gets disabled by 2FA)
- Both features can only be enabled/disabled after an extra password check, made possible
  by a new "identitycheck" feature.
- Adapt user profile and portal to allow managing those new security settings.

rco-odoo commented on odoo/tools/cache.py:
```diff
@@ -28,6 +28,9 @@ def ratio(self):
 
+def lru(method, model):
+    counter = STAT[(model.pool.db_name, model._name, method)]
+    return model.pool._Registry__cache, (model._name, method), counter
```
The explicit mangling makes this code valid again inside a method.

rco-odoo commented on odoo/modules/registry.py:
```diff
@@ -110,6 +110,7 @@ def init(self, db_name):
         self._assertion_report = assertion_report.assertion_report()
         self._fields_by_model = None
         self._post_init_queue = deque()
+        self.__cache = LRU(8192)
```
One should also replace all occurrences of
```python
lazy_property.reset_all(self)
```
by
```python
self.__cache.clear()
```
and remove `lazy_property` from imports.

xmo-odoo replied:
There doesn't seem to be any reset_all on registries, only on envs and models, so this should not be relevant (logically the cache would get cleared but never nuked entirely and re-created from scratch).

xmo-odoo commented on odoo/addons/base/models/res_users.py:
```diff
@@ -1417,3 +1419,115 @@ def change_password_button(self):
+    @api.depends('api_keys_only_explicit')
+    def _compute_keys_only(self):
+        for u in self:
+            u.api_keys_only |= u.api_keys_only_explicit
```
FIXME: should be a straight assignment due to v13 changes.

xmo-odoo commented on odoo/addons/base/models/res_users.py:
```diff
@@ -1417,3 +1419,115 @@ def change_password_button(self):
+            id serial primary key,
+            name varchar not null,
+            user_id integer not null REFERENCES res_users(id),
+            index varchar not null CHECK (char_length(index) = {index_size}),
```
FIXME: use CHARACTER(index_size)?

xmo-odoo commented on odoo/addons/base/models/res_users.py:
```diff
@@ -1417,3 +1419,115 @@ def change_password_button(self):
+        return super().unlink()
+
+class APIKeyDescription(models.TransientModel):
+    _name = _description = 'res.users.apikeys.description'
```
FIXME: now needs ACL, not clear what exactly? /cc @mart-e

xmo-odoo commented on odoo/http.py:
```diff
@@ -1012,18 +1011,29 @@ def authenticate(self, db, login=None, password=None):
+        """
+        self.rotate = True
+        request.uid = self.uid = self.pop('pre_uid')
+        self.session_token = self.pop('pre_session_token')
```
TODO: should we regenerate the session token (and check that it matches the pre-session) instead of just copying over? /cc @tbe-odoo

tbe-odoo replied:
Hello, I think it is safer to regen the session token, someone might override the `_get_session_token_fields` in `res.users` and add a field that potentially changes during the full authentication.

xmo-odoo replied:
Should I also check that the old and new session tokens still match or does that not matter?

tbe-odoo replied:
I don't think we need to check the old one, they shouldn't be different. So your actual solution is working  (if none of the session token fields change) but I think it is preferable to generate a new one in case of a change in-between.

Here is a possible case.
A piece of code decide to change a `session token field` and update the `session_token` value in the session to avoid a session invalidation.
If we use the `pre_session_token` which isn't updated, the session would be broken and the user might be unable to authenticate.

So checking the old session token might not be a good solution because this would invalidate the session which is not the intended behavior by the developer.
So I would generate a new session token and let the next request decide if the session is still valid

xmo-odoo replied:
in that case the pre_session_token doesn't seem very useful? I was storing it because it's the one generated during the first phase of the login and so that's the one for which we'd verified the login / password, but if we're not keeping it I can just remove the entire thing.

tbe-odoo replied:
Yup I think we can use the `session_token` directly

xmo-odoo commented on addons/auth_totp/controllers/home.py:
```diff
@@ -0,0 +1,32 @@
+                    user._totp_check(int(kwargs['totp_token']))
+                    request.session.finalize()
+                    return http.redirect_with_hash(self._login_redirect(request.session.uid, redirect=redirect))
+            except AccessDenied:
```
FIXME: also check valueerror in case user passes totp_token=qrux or something?

xmo-odoo commented on odoo/addons/base/security/ir.model.access.csv:
```diff
@@ -68,6 +68,7 @@
 "access_res_users_all","res_users all","model_res_users",,1,0,0,0
 "access_res_users_group_erp_manager","res_users group_erp_manager","model_res_users","group_erp_manager",1,1,1,1
 "access_res_users_log_all","res_users_log_all","model_res_users_log",,1,0,1,0
+"access_res_users_apikeys_access_employee","API keys access","model_res_users_apikeys","group_user",1,0,0,1
```
FIXME: what about portal users?

xmo-odoo commented on odoo/addons/base/models/res_users.py:
```diff
@@ -1427,3 +1419,115 @@ def change_password_button(self):
+
+    name = fields.Char("Description", required=True)
+
+    def make_key(self):
```
FIXME: how do portal users create API keys? Should they even be able to create / have API keys? if not, why do I have an ir.rule for that? /cc @odony

xmo-odoo commented on odoo/addons/base/models/res_users.py:
```diff
@@ -1427,3 +1423,116 @@ def change_password_button(self):
+        INSERT INTO res_users_apikeys (name, user_id, key, index)
+        VALUES (%s, %s, %s, %s)
+        RETURNING id
+        """, [self.sudo().name, self.env.user.id, hash_api_key(k), k[:INDEX_SIZE]])
```
@odony I'm hashing the hexlified version of the key here rather than the actual binary data, should not make any real difference and avoids having to unhexlify it on verification but it's kinda weird, what's your take?

tbe-odoo commented on odoo/service/security.py:
```diff
@@ -24,5 +17,5 @@ def check_session(session, env):
     expected = self._compute_session_token(session.sid)
     if expected and odoo.tools.misc.consteq(expected, session.session_token):
```
Shouldn't we check the `pre_session_token` if the MFA url is set and the `session_token` if not?

xmo-odoo replied:
No, we don't want to risk a "pre-session" to be considered a valid session for any situation other than submitting an MFA validation code.

xmo-odoo commented on addons/auth_totp/static/src/js/totp_common.js:
```diff
@@ -0,0 +1,66 @@
+        var $input = this.$('input[name=code]');
+        var code = Number($input.val())|0;
+
+        console.log('tryEnabling', code);
```
delete this

xmo-odoo commented on odoo/addons/base/views/res_users_views.xml:
```diff
@@ -336,6 +336,60 @@
+                        <p><field name="key" readonly="True"/></p>
+                        <p class="bg-warning">
+                            This key will not be re-issued, and provides all
+                            access to your user account, it is very important to
```
"all" -> "full", and bold it.

mart-e commented on addons/auth_totp/models/res_users.py:
```diff
@@ -0,0 +1,142 @@
+    def match(self, code, *, previous, t=None, window=TIMESTEP):
+        """
+        :param code: authenticator code to check against this key
+        :param int previous: counter value returned by a prevously successful
```
previously

mart-e commented on addons/auth_totp/views/templates.xml:
```diff
@@ -0,0 +1,41 @@
+                <form method="POST" action="/web/session/logout" class="form-inline">
+                    <input type="hidden" name="csrf_token" t-att-value="request.csrf_token()"/>
+                    <button type="submit" class="btn btn-link btn-sm mb-2">
+                        Cancel and log out
```
while I understand the intention, it is a bit odd to say "and log out", the user is trying to log in. By saying this, you imply they are already logged in.

mart-e commented on addons/auth_totp/static/src/xml/templates.xml:
```diff
@@ -0,0 +1,12 @@
+        </p>
+        <!-- provide secret string or explicit link? -->
+
+        <p><label for="code">Input authenticator-provided code: <input name="code" class="o_input o_required_modifier" autofocus="autofocus"/></label></p>
```
missing a bit of clarification what this field is, maybe something like "Enter the generated authenticator-provided code to confirm the activation of ..."

xmo-odoo commented:
There, should be about OK. Since last time:

* added a "security" portal page so portal users can change their passwords, and create API keys, and enable totp
* the totp UI has been completely rewritten to be a standard workflow which the frontend reuses / emulates / massages, that way they can stay in sync but the backend flow is *normal* instead of some weird-ass custom dialog mess (the frontend is still a weird-ass custom dialog mess)
* also rebuilt the LRU on ordereddict as that seemed like a pretty good idea at the time (and building LRU is a big part of ordereddict's purpose)

xmo-odoo commented:
ping @aab-odoo (for the JS stuff) @qsm-odoo (for the styling / boostrap / frontend)

xmo-odoo commented:
@odony 

> check password when enabling/disabling TOTP and API Keys

How would that work for auth schemes which are not password-based e.g. oauth or SSO? Duplicating the login flow seems difficult.

odony commented:
> How would that work for auth schemes which are not password-based e.g. oauth or SSO? Duplicating the login flow seems difficult.

Is there any case besides auth_oauth? I don't think MFA + OAuth makes a lot of sense, the MFA should be on the OAuth Authority side. Same thing for the password change, it won't work for OAuth.
If we find a way to detect it, we should probably disable all those security features for OAuth sessions. Perhaps we could flag the sessions at login?

For LDAP I think it's ok, the password change will work (soon) and the password check will work fine as well.

xmo-odoo commented:
> I don't think MFA + OAuth makes a lot of sense

No but api keys + oauth does.

xmo-odoo commented:
@odony that aside I've fixed the display of the portal totp dialog on chrome, added a link (seems to work on ios at least) & a revealable display of the secret itself, & moved the creation of a portal API key behind a dialog.

odony commented:
> No but api keys + oauth does.

True, but it's not a major focus. I think we can ignore that for now.
Creating a local user dedicated to RPC is always a workaround for those odd cases (with the assistance of an admin). It's API keys in a slightly more manual way ;-)

xmo-odoo commented:
@odony password protection thing should work, also added a tour because @aab-odoo threatened me (and fixed some stuff in the browser running as it had issues with using multiple tours as different users in the same test / test class).

Julien00859 commented on odoo/addons/base/models/res_users.py:
```diff
@@ -1381,6 +1416,29 @@ def fields_get(self, allfields=None, attributes=None):
+        if not request:
+            raise UserError(_("This method can only be accessed over HTTP"))
+        self.create_uid._check_credentials(self.password, {'interactive': True})
+        self.password = False
```
shouldn't we flush here ?

xmo-odoo replied:
No? It should automatically be flushed at the first read or when the query ends, the point is just to "nuke" the password from this transient as the transient has been "consumed" by the check.

Julien00859 commented:
> * hash keys just in case as we can do so and might as well, add a
  cleartext index (first 4 bytes of 20) to avoid blowing up the DB if
  a user decides to create millions of keys for some daft reason

Hello xmo, can you elaborate this a bit ? Would the index "just" be 5x smaller or is there anything else ?

xmo-odoo commented:
> Hello xmo, can you elaborate this a bit ? Would the index "just" be 5x smaller or is there anything else ?

It's a cleartext index to hashed API keys. Without an index, in order to check keys we'd need a linear search through the table, which would / could be an issue if one user created *a lot* of keys. 

The validation takes about 5ms per key, so if a user created a million of API keys going through all of them would take an hour.

xmo-odoo commented:
ping @odony

amh-mw commented on addons/portal/static/src/xml/portal_security.xml:
```diff
@@ -0,0 +1,46 @@
+                Here is your new API key, use it instead of a
+                password when doing RPC access. Your login is still
+                necessary.
+            </p>
```
I have some concern about the capability (and desire) of users to understand terms like API and RPC. For our own purposes, we have begun discussing some way to bootstrap this process in our iOS app (captive browser? main browser launches link into our app?) such that the user enters their password (and 2FA) to confirm this action, then the API key is stored in the keychain on the device (requiring FaceId/TouchId?) and is never really seen by the user.

xmo-odoo replied:
> I have some concern about the capability (and desire) of users to understand terms like API and RPC. 

The entire point of that wizard is to generate keys *for RPC access*. 

> For our own purposes, we have begun discussing some way to bootstrap this process in our iOS app (captive browser? main browser launches link into our app?) such that the user enters their password (and 2FA) to confirm this action, then the API key is stored in the keychain on the device (requiring FaceId/TouchId?) and is never really seen by the user.

Yes, you can reuse the idea and storage without using the same UI. I'm pretty sure that's what @Florimond does as their concern is oauth access from / for third-party system, so the user never sees the key. They probably edited some bits in order to make this non-interactive flow more convenient.

xmo-odoo commented on addons/auth_totp/models/res_users.py:
```diff
@@ -154,13 +151,14 @@ class TOTP:
         self._key = key
 
-    def match(self, code, *, previous, t=None, window=TIMESTEP):
+    def match(self, code, previous=0, t=None, window=TIMESTEP):
```
What's the point of leaving the parameter if you've removed the feature?

xmo-odoo commented on addons/auth_totp/models/res_users.py:
```diff
@@ -99,7 +93,6 @@ def totp_enable_wizard(self):
 
 class TOTPWizard(models.TransientModel):
     _name = 'auth_totp.wizard'
-    _description = "wizard used to set the totp token on a user's account"
```
not going to work, descriptions are pretty much required now.

xmo-odoo commented on addons/auth_totp/models/res_users.py:
```diff
@@ -18,9 +18,8 @@
-    totp_last_valid = fields.Integer(default=0)
-    totp_enabled = fields.Boolean(string="TOTP enabled", compute='_compute_totp_enabled', store=True)
+    totp_secret = fields.Char(copy=False, groups=".") # no access
+    totp_enabled = fields.Boolean(string="TOTP enabled", compute='_compute_totp_enabled')
```
why unstore the information? why remove the replay protection?

xmo-odoo commented on odoo/addons/base/models/res_users.py:
```diff
@@ -1456,3 +1523,148 @@ def change_password_button(self):
+            except AccessDenied:
+                pass
+
+        if self.env['res.users.apikeys']._check_credentials(scope='rpc', key=password) == self.env.uid:
```
@odony used "rpc" for the scope of the key-based authentication, I think it doesn't quite match but I can't find a better scope name.

hbto commented:
It is wonderfull to see this coming to Odoo.

I would like to point out that seed to validate the 2FA OTP shall not be saved inside same DB.
They have to be saved elsewhere.

Otherwise you are putting the fox and the chickens in the same barnyard

Regards.

nhomar commented:
![Finally](https://media.giphy.com/media/QwnbJC6EMTjDJ7la8T/giphy.gif)

nhomar commented:
This is an amazing job.

I would love you try the following scenarios (we use them all).

1. Logged with oauth (gsuite or office365 or even odoo.com) + 2FA
2. Use token in parallel.
3. Logged with Active directory + 2FA.


Today when being logged with 2 different methods (let's say in one windows with oauth and other with password) the system start to crash asking me to log me in again over and over again.

Do to this refactor this should be a nice moment to check such use cases.

But it is AMAZING we will have 2020 authentication methods available finally !

Kudos team!

moylop260 commented on odoo/service/security.py:
```diff
@@ -24,5 +17,5 @@ def check_session(session, env):
     if expected and odoo.tools.misc.consteq(expected, session.session_token):
         return True
-    self._invalidate_session_cache()
+    self.clear_caches()
```
@xmo-odoo 

@guewen @gurneyalex @odony and I have participated in the following thread:
 - https://twitter.com/guewenb/status/1278946522068123650?s=21

The global cache is cleared in all workers for all models.
Is not this line execute too frequently?

xmo-odoo replied:
@moylop260 it should be OK: this should only be reached when a user has a session which still exists but whose token is invalid, and once `check)sessuib` returns `False` the session is immediately destroyed and this is not even reachable (with the same session-id). Though it's possible that the cache-clearing is unnecessary regardless, we need to check why it was adde with the original author and (eventually) whether that's still needed.

oscarolar commented:
Hello, @xmo-odoo 

I was testing this toy in runbot and I configured the instance to test with Google Oauth + 2FA and is not working.

are you planning on supporting it?

regards.

odony commented:
> I was testing this toy in runbot and I configured the instance to test with Google Oauth + 2FA and is not working.
> 
> are you planning on supporting it?

It _should_ be working after my last changes. However the setup will be awkward, as enabling 2FA requires confirming your password first, and with OAuth you only have a token. You'll need to set a local password temporarily for that, or extract the oauth token to use it as the password. Exercise left to the reader - one would think that using 2FA on the OAuth provider side would be more practical anyway.

odony commented:
> It is wonderfull to see this coming to Odoo.

Yep, and I'm to blame because @xmo-odoo has this thing sitting almost ready for a while! 

> I would like to point out that seed to validate the 2FA OTP shall not be saved inside same DB.
> They have to be saved elsewhere.

Well, for the server-side it's quite practical to have the TOTP secrets in the same database. You could harden it by using secure enclaves and whatnot, but that means it will be harder to deploy, and therefore less likely to be done. More is less  :p

Same for users. In theory you'd think that it's not good to store your password and your TOTP secret in the same password manager on the same device. But in reality it's *fine*. It means it will be convenient so you will use it a lot. Sure, you'd get better theoretical security if you kept these things in separate places. But in practice it will be worse because you won't do it or you will cheat. Another example: password policies from hell. And the list goes on...

We just have to wait for technology to bring more convenience (password managers, biometric smartphones, etc) and then the level of security raises naturally :-)

odony commented:
@xmo-odoo So I think I'm done with my review, here's a summary of the main changes I made, as discussed:

- de-stored field "totp_enabled", does not seem useful, 1 less column on res.users 🙏. I don't think we want anybody to be able to search/group on that, it's a per-user setting.
- removed stored field "totp_last_valid". Made the `previous` arg of `hotp` useless, so dropped it. The purpose was noble but it does not seem to provide a tremendous amount of extra security, given the short life of tokens with the default window. It also costs an extra field on res.users, and a mandatory UPDATE on the user row *every time* the user performs a login, i.e. an exclusive lock on a row that is pointed by *a lot* of FKs (create_uid/write_uid). This is why we had to move the "login_date" outside of res.users too. You'd think that the PG 9.3 lock mode "FOR NO KEY UPDATE" would solve this, but there are still broken cases, at least up to 9.5 and most likely even later, due to the serializable mode heuristics. Cfr https://github.com/odoo/odoo/issues/8585#issuecomment-145926581 It's just really much better to avoid that if we can, and I don't think this feature is worth an extra table.
- prettified a bit the various backend wizard forms ("enhanced security" / "password check" + "totp wizard")
- moved all password & auth-related settings into a new "Account Security" page in user profile form. 
- removed empty auth_totp_hr completely, not necessary anymore as the notebook is integrated and extended into the final HR profile form.
- some security-related changes:
  - `totp_secret` was readable by anyone. Adding it to 'SELF_READABLE_FIELDS' does not protect it from being accessed by other users.
  - totp_try_setting() did not need to be public
  - ir.rule on totp wizard can be global, for extra safety by default
  - forced "secret" to NULL on wizard after TOTP setup, because why keep it
- removed the "API Keys" exposition on portal. Portal users who ever need that can request a key from an admin.
- added LOGGING for each totp operation
- added LOGGING for API key add/remove
- documented our TOTP secret size wrt to RFC4226,4.R6. Google Auth and many vendors (incl. Github) use 80 bits, against the recommendation of the RFC.
- renamed some of the user-facing "TOTP" refs (also logs), most people have heard of "Two-factor authentication" but few know what "TOTP" is, or what "Secure Login" could mean. The screens will tell them that it's not SMS-based 2FA.
- made it possible for `autocomplete` to be forced on JS password fields -> set autocomplete='current-password' in Identity check form, for proper behavior
- fix: /web/login/totp was accessible with not pre_uid and crashing upon submit
- fix: after changing password, portal user was logged out but did not know it until the next click
- added creation date on API keys, helps remember what they were created for
- made api key date and scope visible in form/list
- added assert to make `scope` required in res.users.apikeys.check_credentials()
- removed unnecessary clear_cache at the end of security.check_session() - the cache is already presumed clean at that point, no need to invalidate twice
- verified checked calls to `session.authenticate()`, as the result is different and the session may only be pre-authenticated:
  - base_module_import: KO -> use API KEY instead of password, it's expected.
  - db_manager creation and auth_signup: OK, 2FA is not present at this stage, behavior unchanged
  - login_and_redirect -> KO and broke OAuth: fixed it
- OAuth + 2FA should now work except that setting up 2FA requires a password check, so a local password initially (or knowledge of the oauth token)
- remove `api_key_only_explicit`, as there seemed to be no real advantage in forbidding password-based RPC outside of the 2FA case (interactive JSON routes can be scripted to expose the same API). Replaced by a simpler method `_rpc_api_keys_only()` overridden by `auth_totp` when TOTP enabled.
- auth_ldap and auth_oauth were allowing RPC access even after activating 2FA - fixed it by taking `_rpc_api_keys_only()` into account there as well
- log_ormcache_stats was broken by the cache change, fixed it

One thing remains in my todo list for this, but can be done later:
  - send email notification on TOTP/API key changes (and ideally other things like password or email change, when done by the user)

oscarolar commented:
> > I was testing this toy in runbot and I configured the instance to test with Google Oauth + 2FA and is not working.
> > are you planning on supporting it?
> 
> It _should_ be working after my last changes. However the setup will be awkward, as enabling 2FA requires confirming your password first, and with OAuth you only have a token. You'll need to set a local password temporarily for that, or extract the oauth token to use it as the password. Exercise left to the reader - one would think that using 2FA on the OAuth provider side would be more practical anyway.

Hello @odony and thanks for addressing the issue, there is a case of use for this scenario with password enabled and Google Auth, I have to enable the password when I use Odoo's mobile app since it doesn't support any oauth providers, on the other hand when I'm on the browser I use the fastest way to log in which is using a 3rd Party auth.

Does it make sense to you?

odony commented on odoo/service/security.py:
```diff
@@ -24,5 +17,4 @@ def check_session(session, env):
     expected = self._compute_session_token(session.sid)
     if expected and odoo.tools.misc.consteq(expected, session.session_token):
         return True
-    self._invalidate_session_cache()
```
As discussed on a previous version of the commit, at that point in the `check_session()` the cache is supposed to be valid, there is no reason to invalidate it, so we can remove this invalidation.
It only happened when there was a mismatch, so the session had a UID but the token did not match the expected token.
- Either we just had a cache miss for the token, so we'd just refreshed it with the new value, nothing to invalidate
- Or we had a cache hit, so there's perhaps another session for the same user that has seeded the cache. But the cached valued must be good because we invalidate and propagate the invalidation whenever any user attribute changes (cfr `res.users.write()`)

robodoo commented:
Merge method set to rebase and merge, using the PR as merge commit message

odony commented:
> Hello @odony and thanks for addressing the issue, there is a case of use for this scenario with password enabled and Google Auth, I have to enable the password when I use Odoo's mobile app since it doesn't support any oauth providers, on the other hand when I'm on the browser I use the fastest way to log in which is using a 3rd Party auth.
> 
> Does it make sense to you?

Oh sure, so you actually meant OAuth + local password + 2FA, makes more sense indeed. That should work fine now, but you will get the 2FA prompt after the Google Auth, which you may or may not like. I took the point of view where an activated 2FA means you *have to* pass a 2FA prompt at every interactive login. This way it's consistent everywhere (local, LDAP, OAuth).

Maybe @adr-odoo has OAuth2 support in the backlog for the Mobile app, though ;-)

robodoo commented:
Linked pull request(s) odoo/enterprise#9112 not ready. Linked PRs are not staged until all of them are ready.

odony commented:
> Today when being logged with 2 different methods (let's say in one windows with oauth and other with password) the system start to crash asking me to log me in again over and over again.

That should work more or less. I suspect the problem happens when you connect last with OAuth. This updates the access token on the user, so it invalidates all other sessions. If you connect with OAuth first, you can then connect with LDAP/Local as many times as you want, it should not disconnect you. 2FA will not change that, it's a consequence of the way we must invalidate sessions when one of the security properties change.

robodoo commented:
Staging failed: timed out (>120 minutes)

mart-e commented:
robodoo retry

Yenthe666 commented:
This is great work, LOVE it. 👏 
What I can't seem to directly figure out is if it is possible to create tokens with scopes for specific models? E.g generate a token for some user it's API's that can only do CRUD operations on contacts & sale orders.

xmo-odoo commented:
> What I can't seem to directly figure out is if it is possible to create tokens with scopes for specific models? E.g generate a token for some user it's API's that can only do CRUD operations on contacts & sale orders.

That's not currently supported.

"Slicing" existing ACL would be complicated as the inter-weavings are anything but trivial and there was no real use-case for it, at least as of the initial version (the original case for API keys was plain and simply that RPC via password would and should not work on an account with 2FA enabled so there needed to be an alternate mechanism), so the current case is:

* the keys created "by hand" for RPC don't have a scope set and using them over RPC has no restriction aside from normal user ACL
* APIs are available to use custom scopes in custom endpoints, such that the keys for those scopes will not be usable for RPC, and those custom endpoints will explicitly provide whatever access or restrictions they want

Yenthe666 commented:
Okay got it, thanks for the detailed info. So in short: the token follows the same security rules as the ACL's for the user and cannot be deviated from (unless custom code).

xmo-odoo commented:
> Okay got it, thanks for the detailed info. So in short: the token follows the same security rules as the ACL's for the user and cannot be deviated from (unless custom code).

Yup, it's really an alternative means of identification for programmatic access (alternative in the sense that you can use keys even without 2FA active, but have to when 2FA is active on the account), it doesn't use a separate security model.
